File: clients/python/tests/chunking/resources/php/sample.php
<?php
/**
 * Sample PHP file for testing code chunking.
 *
 * This file contains various PHP constructs to test AST parsing.
 */

declare(strict_types=1);

namespace Sample;

use Exception;
use InvalidArgumentException;

// Constants
define('MAX_RETRIES', 3);
const DEFAULT_TIMEOUT = 30.0;

/**
 * Custom exception for service errors.
 */
class ServiceException extends Exception
{
    private ?string $errorCode;

    public function __construct(string $message, ?string $errorCode = null, int $code = 0, ?Exception $previous = null)
    {
        parent::__construct($message, $code, $previous);
        $this->errorCode = $errorCode;
    }

    public function getErrorCode(): ?string
    {
        return $this->errorCode;
    }
}

/**
 * Represents a user in the system.
 */
class User
{
    private string $id;
    private string $name;
    private ?string $email;

    public function __construct(string $id, string $name, ?string $email = null)
    {
        $this->id = $id;
        $this->name = $name;
        $this->email = $email;
    }

    public function getId(): string
    {
        return $this->id;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    public function setEmail(string $email): void
    {
        $this->email = $email;
    }

    public function getDisplayName(): string
    {
        return $this->name ?: ($this->email ?: $this->id);
    }

    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'email' => $this->email,
        ];
    }
}

/**
 * Interface for services.
 */
interface ServiceInterface
{
    public function initialize(): void;
    public function isReady(): bool;
}

/**
 * Trait for common functionality.
 */
trait Loggable
{
    protected function log(string $message): void
    {
        echo "[LOG] $message\n";
    }
}

/**
 * Base class for services.
 */
abstract class BaseService implements ServiceInterface
{
    use Loggable;

    protected array $config;
    protected bool $initialized = false;

    public function __construct(array $config = [])
    {
        $this->config = $config;
    }

    public function initialize(): void
    {
        $this->initialized = true;
    }

    public function isReady(): bool
    {
        return $this->initialized;
    }

    protected function validateConfig(): bool
    {
        return !empty($this->config);
    }
}

/**
 * Service for managing users.
 */
class UserService extends BaseService
{
    private array $users = [];

    public function createUser(string $id, string $name, ?string $email = null): User
    {
        if (empty($id)) {
            throw new ServiceException('ID cannot be empty', 'INVALID_ID');
        }

        $user = new User($id, $name, $email);
        $this->users[$id] = $user;
        $this->onUserCreated($user);
        return $user;
    }

    public function getUser(string $id): ?User
    {
        return $this->users[$id] ?? null;
    }

    public function deleteUser(string $id): bool
    {
        if (isset($this->users[$id])) {
            unset($this->users[$id]);
            return true;
        }
        return false;
    }

    public function getUsers(): array
    {
        return array_values($this->users);
    }

    private function onUserCreated(User $user): void
    {
        // Internal callback
    }
}

/**
 * Calculate factorial of n.
 */
function calculateFactorial(int $n): int
{
    if ($n <= 1) {
        return 1;
    }
    return $n * calculateFactorial($n - 1);
}

/**
 * Fetch data from URL (simulated).
 */
function fetchData(string $url, float $timeout = DEFAULT_TIMEOUT): array
{
    // Simulated fetch
    return [
        'url' => $url,
        'status' => 'ok',
        'timeout' => $timeout,
    ];
}

/**
 * Process items with optional validation.
 */
function processItems(array $items, bool $validate = true): array
{
    if ($validate) {
        $items = array_filter($items, fn($item) => !empty($item));
    }
    return array_map(fn($item) => strtolower(trim($item)), $items);
}

// Arrow function (PHP 7.4+)
$doubleValue = fn(int $x): int => $x * 2;

// Main execution
if (php_sapi_name() === 'cli' && basename(__FILE__) === basename($argv[0] ?? '')) {
    $service = new UserService(['env' => 'test']);
    $service->initialize();

    $user = $service->createUser('1', 'Test User', 'test@example.com');
    echo "Created user: " . $user->getDisplayName() . "\n";

    $result = calculateFactorial(5);
    echo "Factorial: $result\n";
}

File: clients/python/tests/chunking/resources/perl/sample.pl
#!/usr/bin/perl
#
# Sample Perl file for testing code chunking.
#
# This file contains various Perl constructs to test AST parsing.

use strict;
use warnings;
use feature 'say';
use Carp qw(croak);

# Constants
use constant MAX_RETRIES => 3;
use constant DEFAULT_TIMEOUT => 30.0;

# Package for User
package User;

sub new {
    my ($class, %args) = @_;
    croak "ID is required" unless defined $args{id};

    my $self = {
        id    => $args{id},
        name  => $args{name} // '',
        email => $args{email},
    };

    return bless $self, $class;
}

sub id { shift->{id} }

sub name {
    my ($self, $value) = @_;
    if (defined $value) {
        $self->{name} = $value;
    }
    return $self->{name};
}

sub email {
    my ($self, $value) = @_;
    if (defined $value) {
        $self->{email} = $value;
    }
    return $self->{email};
}

sub get_display_name {
    my $self = shift;
    return $self->{name} || $self->{email} || $self->{id};
}

sub to_hash {
    my $self = shift;
    return {
        id    => $self->{id},
        name  => $self->{name},
        email => $self->{email},
    };
}

# Package for UserService
package UserService;

sub new {
    my ($class, %args) = @_;

    my $self = {
        users       => {},
        initialized => 0,
        config      => $args{config} // {},
    };

    return bless $self, $class;
}

sub initialize {
    my $self = shift;
    $self->{initialized} = 1;
    return 1;
}

sub is_ready {
    my $self = shift;
    return $self->{initialized};
}

sub create_user {
    my ($self, %args) = @_;

    croak "ID cannot be empty" unless defined $args{id} && $args{id} ne '';

    my $user = User->new(%args);
    $self->{users}{$args{id}} = $user;
    $self->_on_user_created($user);

    return $user;
}

sub get_user {
    my ($self, $id) = @_;
    return $self->{users}{$id};
}

sub delete_user {
    my ($self, $id) = @_;

    if (exists $self->{users}{$id}) {
        delete $self->{users}{$id};
        return 1;
    }
    return 0;
}

sub get_all_users {
    my $self = shift;
    return values %{$self->{users}};
}

sub _on_user_created {
    my ($self, $user) = @_;
    # Internal callback
}

# Back to main package
package main;

# Calculate factorial of n
sub calculate_factorial {
    my $n = shift;
    return 1 if $n <= 1;
    return $n * calculate_factorial($n - 1);
}

# Fetch data from URL (simulated)
sub fetch_data {
    my ($url, $timeout) = @_;
    $timeout //= DEFAULT_TIMEOUT;

    return {
        url     => $url,
        status  => 'ok',
        timeout => $timeout,
    };
}

# Process items with optional validation
sub process_items {
    my ($items, $validate) = @_;
    $validate //= 1;

    my @filtered = $validate
        ? grep { defined $_ && $_ ne '' } @$items
        : @$items;

    return [ map { lc(trim($_)) } @filtered ];
}

# Helper to trim whitespace
sub trim {
    my $str = shift;
    $str =~ s/^\s+|\s+$//g;
    return $str;
}

# Retry a block with exponential backoff
sub with_retry {
    my ($max_retries, $block) = @_;
    $max_retries //= MAX_RETRIES;

    my $retry = 0;
    while ($retry < $max_retries) {
        my $result = eval { $block->() };
        return $result unless $@;
        $retry++;
        sleep(2 ** $retry);
    }
    die "Max retries exceeded: $@";
}

# Main execution
sub main {
    my $service = UserService->new(config => { env => 'test' });
    $service->initialize();

    my $user = $service->create_user(
        id    => '1',
        name  => 'Test User',
        email => 'test@example.com'
    );
    say "Created user: " . $user->get_display_name();

    my $result = calculate_factorial(5);
    say "Factorial: $result";
}

# Run main if executed directly
main() if !caller();

1;  # Return true for module
